snuba/query/expressions.py

```diff
@@ -110,10 +110,88 @@ def visit_lambda(self, exp: Lambda) -> TVisited:
         raise NotImplementedError
 
 
+class StringifyVisitor(ExpressionVisitor[str]):
+    """Visitor implementation to turn an expression into a string format
+
+    Usage:
+        # Any expression class supported by the visitor will do
+        >>> exp: Expression = Expression()
+        >>> visitor = StringifyVisitor()
+        >>> exp_str = exp.accept(visitor)
+    """
+
+    def __init__(self) -> None:
+        # keeps track of the level of the AST we are currently in, this is necessary for nice indentation
+        # before recursively going into subnodes increment this counter, decrement it after the recursion
+        # is done
+        self.__level = 0
+
+    def _get_line_prefix(self) -> str:
+        # every line in the tree needs to be indented based on the tree level
+        # to make things look pretty
+        return "\n" + "  " * self.__level
+
+    def _get_alias_str(self, exp: Expression) -> str:
+        # Every expression has an optional alias so we handle that here
+        return f" AS `{exp.alias}`" if exp.alias else ""
+
+    def visit_literal(self, exp: Literal) -> str:
+        literal_str = None
+        if isinstance(exp.value, str):
+            literal_str = f"'{exp.value}'"
+        elif isinstance(exp.value, datetime):
+            literal_str = f"datetime({exp.value.isoformat()})"
+        elif isinstance(exp.value, date):
+            literal_str = f"date({exp.value.isoformat()})"
+        else:
+            literal_str = f"{exp.value}"
+        res = f"{self._get_line_prefix()}{literal_str}{self._get_alias_str(exp)}"
+        return res
+
+    def visit_column(self, exp: Column) -> str:
+        column_str = (
+            f"{exp.table_name}.{exp.column_name}"
+            if exp.table_name
+            else f"{exp.column_name}"
+        )
+        return f"{self._get_line_prefix()}{column_str}{self._get_alias_str(exp)}"
+
+    def visit_subscriptable_reference(self, exp: SubscriptableReference) -> str:
+        # this line will already have the necessary prefix due to the visit_column
+        # function
+        subscripted_column_str = f"{exp.column.accept(self)}['{exp.key.value}']"
```

Comment:
Being key a literal shouldn't it accept the visitor as well ? In practice they are always strings and never have an alias but this is not guaranteed.

Reply:
I was trying to be cheeky but you're right I shouldn't. Will do